### ISSUE TITLE
Crashing for Python3 list contractions over statically known to be non-iterable value 

### ISSUE THREAD

**Issue by gggggeorge** [Nuitka/Nuitka#513]
Hello,

I just started using Nuitka today, and it is amazing! I successfully created some EXEs for a few Python scripts, but when I tried to use it to build an EXE for a python GUI (PySimpleGUIQt), Nuitka seemed to have some problems...?

I have the following Python script (just a single import):

**script_Run-SendFile_GUI-SIMPLE.py**:

`import PySimpleGUIQt as sg`

----
----
This is on Windows 10 (64 bit):

**python -m nuitka --version**

```
0.6.5
Python: 3.5.4 (v3.5.4:3f56838, Aug  8 2017, 02:17:05) [MSC v.1900 64 bit (AMD64)]
Executable: C:\Python35\python.exe
OS: Windows
Arch: x86_64
```

Here is how I installed the Python libraries:

```
pip install PySimpleGUIQt
pip install PySide2
pip install PyQt5
pip install nuitka
```

----
----
Here is what I did to create the EXE:

**python -m nuitka --follow-imports script_Run-SendFile_GUI-SIMPLE.py**

Here is the error message I get:

```
Problem with <Node 'EXPRESSION_FUNCTION_BODY' with {'parameters': <ParameterSpec 'self,event'>, 'name': '_treeview_selected', 'code_object': <CodeObjectSpec Function '_treeview_selected' with 'self,event,selections,MyForm'>, 'flags': None, 'ref_name': 'PySimpleGUIQt$PySimpleGUIQt$$$function_146__treeview_selected', 'provider': 'PySimpleGUIQt$PySimpleGUIQt', 'doc': None}>
Problem with <Node 'COMPILED_PYTHON_MODULE' with {'name': 'PySimpleGUIQt', 'package': 'PySimpleGUIQt', 'filename': 'C:\\Python35\\lib\\site-packages\\PySimpleGUIQt\\PySimpleGUIQt.py'}>
Traceback (most recent call last):
  File "C:\Python35\lib\site-packages\nuitka\__main__.py", line 184, in <module>
    main()
  File "C:\Python35\lib\site-packages\nuitka\__main__.py", line 177, in main
    MainControl.main()
  File "C:\Python35\lib\site-packages\nuitka\MainControl.py", line 759, in main
    main_module = createNodeTree(filename=filename)
  File "C:\Python35\lib\site-packages\nuitka\MainControl.py", line 141, in createNodeTree
    Optimization.optimize(main_module.getOutputFilename())
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 545, in optimize
    finished = makeOptimizationPass(initial_pass=False)
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 474, in makeOptimizationPass
    if optimizeVariables(current_module):
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 370, in optimizeVariables
    if optimizeUnusedUserVariables(function_body):
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 325, in optimizeUnusedUserVariables
    function_body.removeUserVariable(local_variable)
  File "C:\Python35\lib\site-packages\nuitka\nodes\FunctionNodes.py", line 224, in removeUserVariable
    assert variable in self.providing.values(), (self, self.providing, variable)
AssertionError: (<Node 'EXPRESSION_FUNCTION_BODY' with {'parameters': <ParameterSpec 'self,event'>, 'name': '_treeview_selected', 'code_object': <CodeObjectSpec Function '_treeview_selected' with 'self,event,selections,MyForm'>, 'flags': None, 'ref_name': 'PySimpleGUIQt$PySimpleGUIQt$$$function_146__treeview_selected', 'provider': 'PySimpleGUIQt$PySimpleGUIQt', 'doc': None}>, {'self': <ParameterVariable 'self' of '_treeview_selected'>, 'selections': <LocalVariable 'selections' of '_treeview_selected'>, 'event': <ParameterVariable 'event' of '_treeview_selected'>, 'MyForm': <LocalVariable 'MyForm' of '_treeview_selected'>}, <LocalVariable 'x' of 'listcomp'>)
```

Thanks,
George Hynes
09/16/2019 6:03pm ET

**Comment by gggggeorge** [comment]
Hello,

I uninstalled your 0.6.5, then installed your 0.6.6rc5 "developer" version to see if it made any difference -and it appears the issue still exists, as per:

```
pip uninstall nuitka
pip install -U "https://github.com/Nuitka/Nuitka/archive/develop.zip"
```

**python -m nuitka --version**

```
0.6.6rc5
Python: 3.5.4 (v3.5.4:3f56838, Aug  8 2017, 02:17:05) [MSC v.1900 64 bit (AMD64)]
Executable: C:\Python35\python.exe
OS: Windows
Arch: x86_64
```

----
----
**python -m nuitka --follow-imports script_Run-SendFile_GUI-SIMPLE.py**

```
Problem with <Node 'EXPRESSION_FUNCTION_BODY' with {'parameters': <ParameterSpec 'self,event'>, 'name': '_treeview_selected', 'code_object': <CodeObjectSpec Function '_treeview_selected' with 'self,event,selections,MyForm'>, 'flags': None, 'ref_name': 'PySimpleGUIQt$PySimpleGUIQt$$$function_146__treeview_selected', 'provider': 'PySimpleGUIQt$PySimpleGUIQt', 'doc': None}>
Problem with <Node 'COMPILED_PYTHON_MODULE' with {'module_name': <ModuleName PySimpleGUIQt.PySimpleGUIQt>, 'filename': 'C:\\Python35\\lib\\site-packages\\PySimpleGUIQt\\PySimpleGUIQt.py'}>
Traceback (most recent call last):
  File "C:\Python35\lib\site-packages\nuitka\__main__.py", line 184, in <module>
    main()
  File "C:\Python35\lib\site-packages\nuitka\__main__.py", line 177, in main
    MainControl.main()
  File "C:\Python35\lib\site-packages\nuitka\MainControl.py", line 757, in main
    main_module = createNodeTree(filename=filename)
  File "C:\Python35\lib\site-packages\nuitka\MainControl.py", line 141, in createNodeTree
    Optimization.optimize(main_module.getOutputFilename())
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 598, in optimize
    finished = makeOptimizationPass(initial_pass=False)
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 527, in makeOptimizationPass
    if optimizeVariables(current_module):
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 420, in optimizeVariables
    if optimizeUnusedUserVariables(function_body):
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 375, in optimizeUnusedUserVariables
    function_body.removeUserVariable(local_variable)
  File "C:\Python35\lib\site-packages\nuitka\nodes\FunctionNodes.py", line 227, in removeUserVariable
    assert variable in self.providing.values(), (self, self.providing, variable)
AssertionError: (<Node 'EXPRESSION_FUNCTION_BODY' with {'parameters': <ParameterSpec 'self,event'>, 'name': '_treeview_selected', 'code_object': <CodeObjectSpec Function '_treeview_selected' with 'self,event,selections,MyForm'>, 'flags': None, 'ref_name': 'PySimpleGUIQt$PySimpleGUIQt$$$function_146__treeview_selected', 'provider': 'PySimpleGUIQt$PySimpleGUIQt', 'doc': None}>, {'self': <ParameterVariable 'self' of '_treeview_selected'>, 'selections': <LocalVariable 'selections' of '_treeview_selected'>, 'event': <ParameterVariable 'event' of '_treeview_selected'>, 'MyForm': <LocalVariable 'MyForm' of '_treeview_selected'>}, <LocalVariable 'x' of 'listcomp'>)
```


Thanks,
George Hynes
09/17/1029 2:28pm ET

**Comment by kayhayen** [comment]
I managed to reproduce this with `python3.7 bin/nuitka-run --recurse-to=PySimpleGUIQt Mini.py` on my machine as well.

**Comment by kayhayen** [comment]
I managed to distill this minimal case out of `PySimpleGUIQt.py`.

```
def f():
    selections = 0
    [x for x in selections]

```

**Comment by kayhayen** [comment]
You might want to inform upstream that this is code is statically raising:

```
    def _treeview_selected(self, event):
        selections = 000000
        self.SelectedRows = [x for x in selections]
```

Nuitka says this:

> Nuitka:DEBUG:m.py:4 : new_raise : Raising for use of 'iter' on type 'int'.

**Comment by kayhayen** [comment]
The issue appears that somehow Nuitka is convinced that for the Python3 list contraction variable, ownership is with the containing function, which is wrong. It normally doesn't cause issues, because these variables are not normally becoming unused, the iteration statically raising is probably the only scenario where that happens.

**Comment by kayhayen** [comment]
Ok, the issue seems to be that function variables and outline variables are handled the same way, but the code should not remove outline variables from their owner, but them outline only. With a change like that, it seems to be working.

I am putting this through some tests before pushing the fix.

**Comment by kayhayen** [comment]
Please check this out: https://nuitka.net/doc/factory.html it should be better.

**Comment by gggggeorge** [comment]
Hi Kay,

thank you very much for the quick bug fix -it now works!

I am running the following:

```
pip uninstall nuitka
pip install -U "https://github.com/Nuitka/Nuitka/archive/develop.zip"
```

**python -m nuitka --version**

```
0.6.6rc5
Python: 3.5.4 (v3.5.4:3f56838, Aug  8 2017, 02:17:05) [MSC v.1900 64 bit (AMD64)]
Executable: C:\Python35\python.exe
OS: Windows
Arch: x86_64
```


Thanks,
George Hynes
09/18/2019 3:19pm ET

**Comment by kayhayen** [comment]
Just a tiny bit strange in that your instructions given from above would only work now that I merged it to develop branch as part of the pre-release rc6 I just made.

**Comment by gggggeorge** [comment]
Your code is self-healing -it's that good! :smile: Hmmmmm... I think I had a typo in my comments (09/18/2019 3:19pm ET):

I indicated:
```
pip uninstall nuitka
pip install -U "https://github.com/Nuitka/Nuitka/archive/develop.zip"
```

-when it should have been (and what I now seem to recall doing):

```
pip uninstall nuitka
pip install -U "https://nuitka.net/doc/factory.html"
```

Thanks,
George Hynes
09/24/2019 2:01pm ET

**Comment by gggggeorge** [comment]
Hello Kay,

everything has been working, but when I tried to create a "--standalone", it fails with AssertionError:

----
This works:
**python -m nuitka --follow-imports test.py**

----
This does not work (neither of them):
**python -m nuitka --standalone test.py**

```
Nuitka:WARNING:Use '--plugin-enable=qt-plugins' for: Inclusion of Qt plugins.
Traceback (most recent call last):
  File "C:\Python35\lib\site-packages\nuitka\__main__.py", line 184, in <module>
    main()
  File "C:\Python35\lib\site-packages\nuitka\__main__.py", line 177, in main
    MainControl.main()
  File "C:\Python35\lib\site-packages\nuitka\MainControl.py", line 766, in main
    main_module = createNodeTree(filename=filename)
  File "C:\Python35\lib\site-packages\nuitka\MainControl.py", line 141, in createNodeTree
    Optimization.optimize(main_module.getOutputFilename())
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 616, in optimize
    makeOptimizationPass(initial_pass=True)
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 530, in makeOptimizationPass
    changed = optimizeModule(current_module)
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 165, in optimizeModule
    optimizeShlibModule(module)
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 160, in optimizeShlibModule
    Plugins.considerImplicitImports(module=module, signal_change=signalChange)
  File "C:\Python35\lib\site-packages\nuitka\plugins\Plugins.py", line 128, in considerImplicitImports
    plugin.considerImplicitImports(module, signal_change)
  File "C:\Python35\lib\site-packages\nuitka\plugins\PluginBase.py", line 173, in considerImplicitImports
    for full_name, required in self.getImplicitImports(module):
  File "C:\Python35\lib\site-packages\nuitka\plugins\standard\ImplicitImports.py", line 770, in getImplicitImports
    for used_module in module.getUsedModules():
  File "C:\Python35\lib\site-packages\nuitka\nodes\ModuleNodes.py", line 899, in getUsedModules
    self._readPyPIFile()
  File "C:\Python35\lib\site-packages\nuitka\nodes\ModuleNodes.py", line 878, in _readPyPIFile
    assert imported.endswith(")")
AssertionError
```

**python -m nuitka --standalone --plugin-enable=qt-plugins test.py**

```
Nuitka:INFO:Injecting plug-in based post load code for module 'PyQt5.QtCore':
Nuitka:INFO:    Setting Qt library path to distribution folder. Need to avoid loading target
Nuitka:INFO:    system Qt plug-ins, which may be from another Qt version.
Nuitka:INFO:Injecting plug-in based pre load code for module 'PyQt5':
Nuitka:INFO:    Adding dist folder to 'PATH' environment variable.
Traceback (most recent call last):
  File "C:\Python35\lib\site-packages\nuitka\__main__.py", line 184, in <module>
    main()
  File "C:\Python35\lib\site-packages\nuitka\__main__.py", line 177, in main
    MainControl.main()
  File "C:\Python35\lib\site-packages\nuitka\MainControl.py", line 766, in main
    main_module = createNodeTree(filename=filename)
  File "C:\Python35\lib\site-packages\nuitka\MainControl.py", line 141, in createNodeTree
    Optimization.optimize(main_module.getOutputFilename())
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 616, in optimize
    makeOptimizationPass(initial_pass=True)
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 530, in makeOptimizationPass
    changed = optimizeModule(current_module)
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 165, in optimizeModule
    optimizeShlibModule(module)
  File "C:\Python35\lib\site-packages\nuitka\optimizations\Optimization.py", line 160, in optimizeShlibModule
    Plugins.considerImplicitImports(module=module, signal_change=signalChange)
  File "C:\Python35\lib\site-packages\nuitka\plugins\Plugins.py", line 128, in considerImplicitImports
    plugin.considerImplicitImports(module, signal_change)
  File "C:\Python35\lib\site-packages\nuitka\plugins\PluginBase.py", line 173, in considerImplicitImports
    for full_name, required in self.getImplicitImports(module):
  File "C:\Python35\lib\site-packages\nuitka\plugins\standard\ImplicitImports.py", line 770, in getImplicitImports
    for used_module in module.getUsedModules():
  File "C:\Python35\lib\site-packages\nuitka\nodes\ModuleNodes.py", line 899, in getUsedModules
    self._readPyPIFile()
  File "C:\Python35\lib\site-packages\nuitka\nodes\ModuleNodes.py", line 878, in _readPyPIFile
    assert imported.endswith(")")
AssertionError
```

----
----
**test.py**:

`import PySimpleGUIQt as sg`

This is on Windows 10 (64 bit):

**python -m nuitka --version**

```
0.6.6rc5
Python: 3.5.4 (v3.5.4:3f56838, Aug  8 2017, 02:17:05) [MSC v.1900 64 bit (AMD64)]
Executable: C:\Python35\python.exe
OS: Windows
Arch: x86_64
```

Here is how I installed the Python libraries:

```
pip install PySimpleGUIQt
pip install PySide2
pip install PyQt5
pip uninstall nuitka
pip install -U "https://nuitka.net/doc/factory.html"
```

Note: the **pip install -U "https://nuitka.net/doc/factory.html"** was done on 09/18/2019.


Thanks,
George Hynes
10/01/2019 4:24pm ET

**Comment by kayhayen** [comment]
Released as 0.6.6
